lib/service.go

```diff
@@ -61,41 +58,83 @@ func makeRequest(url string, method string, params []interface{}) ([]byte, error
 	}
 	req.Header.Add("Content-Type", "application/json")
 
-	resp, err := http.DefaultClient.Do(req)
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	respBody, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return nil, err
 	}
 
-	return ioutil.ReadAll(resp.Body)
+	return parseRPCResponse(respBody)
+}
+
+type rpcResponseContainer struct {
+	url string
+	err error
+	res *rpcResponse
 }
 
 // ForkchoiceUpdatedV1 TODO
 func (m *RelayService) ForkchoiceUpdatedV1(r *http.Request, args *[]interface{}, result *ForkChoiceResponse) error {
 	method := "engine_forkchoiceUpdatedV1"
 	logMethod := m.log.WithField("method", method)
 
-	relayResp, relayErr := makeRequest(m.relayURL, method, *args)
-	bestResponse := relayResp
+	boostPayloadID := make(hexutil.Bytes, 8)
+	rand.Read(boostPayloadID)
 
-	if relayErr != nil {
-		logMethod.WithFields(logrus.Fields{
-			"error":   relayErr,
-			"url":     m.relayURL,
-			"respond": string(relayResp),
-			"method":  method,
-		}).Warn("Could not make request to relay")
-		return fmt.Errorf("relay error: %v", relayErr)
-	}
-	resp, err := parseRPCResponse(bestResponse)
-	if err != nil {
-		logMethod.WithField("err", err).Error("Could not parse response")
-		return err
+	// Call the relay
+	resultC := make(chan *rpcResponseContainer, len(m.relayURLs))
+	for _, url := range m.relayURLs {
+		go func(url string) {
+			res, err := makeRequest(url, method, *args)
+			resultC <- &rpcResponseContainer{url, err, res}
+		}(url)
 	}
 
-	err = json.Unmarshal(resp.Result, result)
-	if err != nil {
-		logMethod.WithField("err", err).Error("Could not unmarshal response", method, err)
-		return err
+	// Collect the responses
+	hasValidResponse := false
+	for i := 0; i < cap(resultC); i++ {
+		res := <-resultC
+
+		// Check for errors
+		if res.err != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.err, "url": res.url}).Error("error making request to relay")
+			continue
+		}
+		if res.res.Error != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.res.Error, "url": res.url}).Warn("error reply from relay")
+			continue
+		}
+
+		// Decode response
+		forkchoiceResponse := new(ForkChoiceResponse)
+		err := json.Unmarshal(res.res.Result, forkchoiceResponse)
+		if err != nil {
+			logMethod.WithFields(logrus.Fields{"error": err, "data": string(res.res.Result)}).Error("Could not unmarshal response")
+			continue
+		}
+
+		if forkchoiceResponse.Status != ForkchoiceStatusValid && forkchoiceResponse.Status != "SUCCESS" && forkchoiceResponse.Status != "" { // SUCCESS is used by mergemock, although it's not in the engine spec (also accept empty status because mergemock)
+			logMethod.WithFields(logrus.Fields{"error": res.err, "url": res.url, "status": forkchoiceResponse.Status}).Warn("status not valid")
+			continue
+		}
+
+		m.store.SetForkchoiceResponse(boostPayloadID.String(), res.url, forkchoiceResponse.PayloadID.String())
```

Comment:
I'd probably check `forkchoiceResponse.PayloadID != nil` just to be safe. In the execution engine spec, you could have the following case:
```
{payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: null} if the payload is deemed VALID and a build process hasn't been started
```

Reply:
Great input, thanks! Updated